rotator/drain.go

```diff
@@ -0,0 +1,579 @@
+)
+
+type DrainOptions struct {
+	// Continue even if there are pods not managed by a ReplicationController, ReplicaSet, Job, DaemonSet or StatefulSet.
+	Force bool
+
+	// Ignore DaemonSet-managed pods.
+	IgnoreDaemonsets bool
+
+	// Period of time in seconds given to each pod to terminate
+	// gracefully.  If negative, the default value specified in the pod
+	// will be used.
+	GracePeriodSeconds int
+
+	// The length of time to wait before giving up on deletion or
+	// eviction.  Zero means infinite.
+	Timeout time.Duration
+
+	// Continue even if there are pods using emptyDir (local data that
+	// will be deleted when the node is drained).
+	DeleteLocalData bool
+
+	// Namespace to filter pods on the node.
+	Namespace string
+
+	// Label selector to filter pods on the node.
+	Selector labels.Selector
+
+	// OnPodDeletedOrEvicted is called when a pod is evicted/deleted; for printing progress output
+	OnPodDeletedOrEvicted func(pod *corev1.Pod, usingEviction bool)
+
+	// SkipWaitForDeleteTimeoutSeconds ignores pods that have a
+	// DeletionTimeStamp > N seconds. It's up to the user to decide when this
+	// option is appropriate; examples include the Node is unready and the pods
+	// won't drain otherwise
+	SkipWaitForDeleteTimeoutSeconds int
+}
+
+type podDelete struct {
+	pod    corev1.Pod
+	status podDeleteStatus
+}
+
+type podDeleteStatus struct {
+	delete  bool
+	reason  string
+	message string
+}
+
+type waitForDeleteParams struct {
+	ctx                             context.Context
+	pods                            []corev1.Pod
+	interval                        time.Duration
+	timeout                         time.Duration
+	usingEviction                   bool
+	getPodFn                        func(string, string) (*corev1.Pod, error)
+	onDoneFn                        func(pod *corev1.Pod, usingEviction bool)
+	globalTimeout                   time.Duration
+	skipWaitForDeleteTimeoutSeconds int
+}
+
+// Takes a pod and returns a bool indicating whether or not to operate on the
+// pod, an optional warning message, and an optional fatal error.
+type podFilter func(corev1.Pod) (include bool, w *warning, f *fatal)
+type warning struct {
+	string
+}
+type fatal struct {
+	string
+}
+
+const (
+	EvictionKind        = "Eviction"
+	EvictionSubresource = "pods/eviction"
+
+	kDaemonsetFatal      = "DaemonSet-managed pods (use IgnoreDaemonsets to ignore)"
+	kDaemonsetWarning    = "Ignoring DaemonSet-managed pods"
+	kLocalStorageFatal   = "Pods with local storage (use DeleteLocalData to override)"
+	kLocalStorageWarning = "Deleting pods with local storage"
+	kUnmanagedFatal      = "Pods not managed by ReplicationController, ReplicaSet, Job, DaemonSet or StatefulSet (use Force to override)"
+	kUnmanagedWarning    = "Deleting pods not managed by ReplicationController, ReplicaSet, Job, DaemonSet or StatefulSet"
+)
+
+func Drain(client kubernetes.Interface, nodes []*corev1.Node, options *DrainOptions) (err error) {
```

Comment:
```suggestion
func Drain(client kubernetes.Interface, nodes []*corev1.Node, options *DrainOptions) error {
```
Hmm, we normally don't do named returns.

Reply:
Fixed ;)